src/_pytest/cacheprovider.py

```diff
@@ -158,6 +158,33 @@ def __init__(self, config):
         self.lastfailed = config.cache.get("cache/lastfailed", {})
         self._previously_failed_count = None
         self._report_status = None
+        self._skipped_files = 0  # count skipped files during collection due to --lf
+
+    def last_failed_paths(self):
+        """Returns a set with all Paths()s of the previously failed nodeids (cached).
+        """
+        result = getattr(self, "_last_failed_paths", None)
+        if result is None:
+            rootpath = Path(self.config.rootdir)
+            result = {rootpath / nodeid.split("::")[0] for nodeid in self.lastfailed}
+            self._last_failed_paths = result
+        return result
+
+    def pytest_ignore_collect(self, path):
+        """
+        Ignore this file path if we are in --lf mode and it is not in the list of
+        previously failed files.
+        """
+        if not self.active or not self.config.getoption("lf") or not path.isfile():
+            return
+        last_failed_paths = self.last_failed_paths()
+        if not last_failed_paths:
+            # if no previous failures, we want to run all items regardless
+            return
+        result = Path(path) not in self.last_failed_paths()
```

Comment:
```suggestion
        result = Path(path) not in last_failed_paths
```

Reply:
Good catch, thanks; done slightly different, I check for `self.lastfailed` on the `if`